### PR TITLE
Initial refactoring

### PR DESCRIPTION
**Pull request by NikolayPianikov**
Have created just for a review

@dtretyakov please review

### CODE REVIEW

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/DefaultEnvironmentVariables.kt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package jetbrains.buildServer.dotnet
+
+import jetbrains.buildServer.runners.CommandLineEnvironmentVariable
+
+interface DefaultEnvironmentVariables {
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/DefaultEnvironmentVariables.kt]: Could we use just `EnvironmentVariables`?

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/arguments/PackArgumentsProvider.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * Provides arguments to dotnet pack command.
+ */
+class PackArgumentsProvider(
+        _parametersService: jetbrains.buildServer.runners.ParametersService,
+        private val _argumentsService: ArgumentsService)
+    : ArgumentsProviderBase(_parametersService, _argumentsService) {
+
+    protected override fun getArgumentStrings(): Sequence<String> {
+        return kotlin.coroutines.experimental.buildSequence {
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/arguments/PackArgumentsProvider.kt]: Could be replaced by `buildSequence { }`

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/arguments/BuildArgumentsProvider.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * Provides arguments to dotnet build command.
+ */
+class BuildArgumentsProvider(
+        _parametersService: ParametersService,
+        private val _argumentsService: ArgumentsService)
+    : ArgumentsProviderBase(_parametersService, _argumentsService) {
+
+    protected override fun getArgumentStrings(): Sequence<String> = buildSequence {
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/arguments/BuildArgumentsProvider.kt]: Everywhere `protected` keyword could be removed

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/DotnetWorkflowComposer.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            DotnetConstants.COMMAND_TEST to _testArgumentsProvider,
+            DotnetConstants.COMMAND_NUGET_PUSH to _nugetPushArgumentsProvider,
+            DotnetConstants.COMMAND_NUGET_DELETE to _nugetDeleteArgumentsProvider)
+
+    override val target: TargetType
+        get() = TargetType.Tool
+
+    override fun compose(context: WorkflowContext, workflow: Workflow): Workflow {
+        if(workflow.commandLines.any()) {
+            throw RunBuildException("This composer should be a root");
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/DotnetWorkflowComposer.kt]: Everywhere `;` could be removed

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package jetbrains.buildServer.runners
+
+import org.apache.xpath.operations.Bool
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt]: Unused import

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+data class CommandLineResult(
+        private val _exitCode: Sequence<Int>,
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt]: When command line result may contains multiple exist codes and outputs?

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/ParametersServiceImpl.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package jetbrains.buildServer.runners
+
+import jetbrains.buildServer.agent.BuildRunnerContext
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/ParametersServiceImpl.kt]: Unused imports

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/ParametersServiceImpl.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import jetbrains.buildServer.runners.ParametersService
+
+class ParametersServiceImpl(
+        private val _buildStepContext: BuildStepContext) : ParametersService {
+
+    override fun tryGetParameter(parameterType: ParameterType, parameterName: String): String? {
+        when(parameterType) {
+            ParameterType.Runner -> return _buildStepContext.runnerContext.runnerParameters[parameterName]
+            ParameterType.Configuration -> return _buildStepContext.runnerContext.configParameters[parameterName]
+            ParameterType.Environment -> return _buildStepContext.runnerContext.getBuildParameters().getEnvironmentVariables()[parameterName]
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/ParametersServiceImpl.kt]: Could be replaced by `_buildStepContext.runnerContext.buildParameters.environmentVariables[parameterName]`

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/PathsServiceImpl.kt [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import java.io.File
+
+class PathsServiceImpl(
+        private final val _buildStepContext: BuildStepContext,
+        private final val _buildAgentConfiguration: BuildAgentConfiguration,
+        private final val _buildAgentConfigurablePaths: BuildAgentConfigurablePaths) : PathsService {
+
+    override fun getPath(pathType: PathType): java.io.File {
+        when(pathType) {
+            PathType.WorkingDirectory -> return _buildStepContext.runnerContext.getWorkingDirectory()
```
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/PathsServiceImpl.kt]: There all `get%Name%()` could be replaced by `%name%`

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt (hunk 6 above)
Comment (dtretyakov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt]: Maybe we could use multiple `CommandLineResult` instances instead?

Comment (NikolayPianikov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt]: The number of exit codes of command result could be 0 or 1? Do you have any idea which type we could use - nullable is not good from my point if view

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/PathsServiceImpl.kt (hunk 9 above)
Comment (NikolayPianikov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/PathsServiceImpl.kt]: done

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/ParametersServiceImpl.kt (hunk 8 above)
Comment (NikolayPianikov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/ParametersServiceImpl.kt]: done

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/ParametersServiceImpl.kt (hunk 7 above)
Comment (NikolayPianikov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/ParametersServiceImpl.kt]: done

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt (hunk 5 above)
Comment (NikolayPianikov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/runners/CommandLineResult.kt]: done

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/DotnetWorkflowComposer.kt (hunk 4 above)
Comment (NikolayPianikov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/DotnetWorkflowComposer.kt]: done

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/arguments/BuildArgumentsProvider.kt (hunk 3 above)
Comment (NikolayPianikov) [inline plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/arguments/BuildArgumentsProvider.kt]: done

**Pull request merged**